File: internal/seckill/seckill.lua
-- seckill.lua
-- 福袋秒杀核心脚本
-- KEYS[1]: 库存 key，如 seckill:stock:1
-- KEYS[2]: 已抢用户集合 key，如 seckill:users:1
-- ARGV[1]: 用户ID
-- ARGV[2]: 每人限购数量

local stock_key = KEYS[1]
local users_key = KEYS[2]
local user_id = ARGV[1]
local limit = tonumber(ARGV[2])

-- 1. 检查用户是否已经抢过
local already = redis.call('SISMEMBER', users_key, user_id)
if already == 1 then
    return -1  -- 用户已抢过，不能重复抢
end

-- 2. 检查库存
local stock = redis.call('GET', stock_key)
if not stock then
    return -2  -- 库存 key 不存在，活动未初始化
end

stock = tonumber(stock)
if stock <= 0 then
    return 0   -- 库存不足
end

-- 3. 原子扣减库存 + 记录用户
redis.call('DECR', stock_key)
redis.call('SADD', users_key, user_id)

return 1  -- 抢购成功

File: internal/seckill/seckill.go
package seckill

import (
	"context"
	_ "embed"
	"fmt"

	"github.com/redis/go-redis/v9"
)

//go:embed seckill.lua
var luaScript string

var seckillSHA string

// LoadScript 在服务启动时调用一次，把脚本预加载到Redis
func LoadScript(rdb *redis.Client) error {
	ctx := context.Background()
	sha, err := rdb.ScriptLoad(ctx, luaScript).Result()
	if err != nil {
		return fmt.Errorf("加载秒杀脚本失败: %w", err)
	}
	seckillSHA = sha
	return nil
}

// DoSeckill 执行秒杀
// activityID: 福袋活动ID
// userID: 用户ID
// 返回: 1=成功, 0=库存不足, -1=已抢过, -2=活动不存在
func DoSeckill(rdb *redis.Client, activityID string, userID string, limit int) (int, error) {
	ctx := context.Background()
	stockKey := fmt.Sprintf("seckill:stock:%s", activityID)
	usersKey := fmt.Sprintf("seckill:users:%s", activityID)

	result, err := rdb.EvalSha(ctx, seckillSHA, []string{stockKey, usersKey}, userID, limit).Int()
	if err != nil {
		return 0, fmt.Errorf("执行秒杀脚本失败: %w", err)
	}
	return result, nil
}
